# Project 4 Problem 1

# Define lexnode

In [10]:
words = ["*a", "*an", "*and", "*apple", "*bat", "*battle", "*banana"]
class LexNode:
  def __init__(self, char,kid=None,parent=None,isLastSame=False,isLast=False,isChild=False):
    self.char = char
    self.children={}
    self.kid = kid
    self.parent=parent
    self.isLastSame=isLastSame
    self.isLast=isLast
    self.isChild=isChild
  
class LexTree:
  def __init__(self,words):
    #pad word with 0 to ensure not run out of index
    max_length=max(len(word) for word in words)
    self.words_len=[len(word) for word in words]
    padded_words = [word.ljust(max_length, "0") for word in words]
    self.words = padded_words
    
    self.word_max_len=max_length
    
  def convert_to_lex(self):

    word_list_lex=[]
    #Use a 2-D list to represent the char in each word as lexNode
    for word in self.words:
      lex_for_each_word=[]
      for character in word:
        char=LexNode(character)
        lex_for_each_word.append(char)
      word_list_lex.append(lex_for_each_word)
    return word_list_lex
  def LexTree_build(self):
    #iterate through each node,set the isLast,isLastsame,isChild attribute
    word_list_lex=self.convert_to_lex()
    for row in range(len(word_list_lex)):
      for column in range(len(word_list_lex[0])):
        
        if word_list_lex[row][column].char=="0":
          break
        elif row==0:
          current_node=word_list_lex[row][column]
          if column==self.words_len[row]-1:
            current_node.isLast=True
          
        elif column==0:
          pass
        else:
          previous_node=word_list_lex[row][column-1]
          current_node=word_list_lex[row][column]
          compare_node=word_list_lex[row-1][column]
          #print(column)
          
          if current_node.char==compare_node.char:
            continue
          else:
            previous_node.isLastSame=True
            parent_node=word_list_lex[row-1][column-1]
            parent_node.kid=previous_node
            previous_node.parent=parent_node
            previous_node.isChild=True
            
            break
    
    #use a 1d array to store the index for each word's isChild=True position
    for row in range(len(word_list_lex)):
      for column in range(len(word_list_lex[0])):
        current_node=word_list_lex[row][column]
        if column==self.words_len[row]-1:
          current_node.isLast=True
    for row in range(len(word_list_lex)):
      for column in range(len(word_list_lex[0])):
        if row>=1 and column>=1:
          previous_node=word_list_lex[row][column-1]
        
          current_node=word_list_lex[row][column]
          parent_node=word_list_lex[row-1][column]
          if parent_node.isLast and current_node.isLastSame:
            previous_node.isChild=True
            current_node.isChild=False
            parent_node=word_list_lex[row-1][column-1]
            
            parent_node.kid=previous_node
            previous_node.parent=parent_node
    c=[0]
    for row in range(len(word_list_lex)):
      for column in range(len(word_list_lex[0])):
        current_node=word_list_lex[row][column]
        if current_node.isChild:
          c.append(column)
    
    d=[]
    for row in range(len(word_list_lex)):
      e=[]
      for column in range(len(word_list_lex[0])):
        current_node=word_list_lex[row][column]
        e.append(current_node.isLast)
      d.append(e)
    for row in range(len(word_list_lex)):
      str=""
      for column in range(len(word_list_lex[0])):
        current_node=word_list_lex[row][column]
        if current_node.char=="0":
            break
        elif column==c[row]:
          if row==0:
            str=str+current_node.char+"--"
          else:
            str=str+"|--"
        elif column<c[row] :
          str=str+"|  "
        elif column>c[row]:
          if not current_node.isLast:
            str=str+current_node.char+"--"
          else:
            str=str+current_node.char 
          
          
          
      print(str)
    
    return word_list_lex,c
def find_parent_row(array):
  parents=[]
  previous_value=None
  for i, value in enumerate(array):
    if value != previous_value:
      if i==0:
        parents.append(i)
      else:
        parents.append(i-1)
      previous_value = value
  return parents

def read_words_from_file(file_path):
    """
    Reads words from a text file, prepends a "*" to each word, and returns them as a list.

    :param file_path: The path to the text file.
    :return: A list of words, each starting with "*".
    """
    words = []
    try:
        with open(file_path, 'r', encoding='utf-8') as file:  # Try reading with UTF-8 encoding
            for line in file:
                words_in_line = line.strip().split()
                words.extend(["*" + word for word in words_in_line])
    except UnicodeDecodeError:
        with open(file_path, 'r', encoding='windows-1252') as file:  # Fallback to Windows-1252 encoding
            for line in file:
                words_in_line = line.strip().split()
                words.extend(["*" + word for word in words_in_line])
    return words

file_path = 'dict_1.txt'
word_list = read_words_from_file(file_path)
#word_list=words
print(word_list)
Tree=LexTree(word_list)
lex_matrix,c=Tree.LexTree_build()
parents_row=find_parent_row(c)

#print(c)
#print(parents_row)
dict_words = []
 
  
    


    

    









UnicodeDecodeError: 'utf-8' codec can't decode byte 0x92 in position 2729: invalid start byte

In [ ]:
!pip install pygtrie


DEPRECATION: Loading egg at c:\python311\lib\site-packages\sigkernel-0.0.1-py3.11-win-amd64.egg is deprecated. pip 23.3 will enforce this behaviour change. A possible replacement is to use pip for package installation..
DEPRECATION: Loading egg at c:\python311\lib\site-packages\tslearn-0.6.1-py3.11.egg is deprecated. pip 23.3 will enforce this behaviour change. A possible replacement is to use pip for package installation..

[notice] A new release of pip is available: 23.2.1 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


# Build lextree

In [ ]:
from collections import Counter
class LexNode:
    def __init__(self, val):
        self.val = val
        self.children = []
        self.property = 0
        self.idx=0
        
   
    def pretty_str(self, level=0):
       
        if self.property==0:
            ret = "--"+self.val
        elif self.property==2:
            ret="--"+self.val+"\n"
        else:
            ret=self.val
        num_children=len(self.children)
        for i in range(num_children):
            if level!=0:
                if num_children>=2 and i!=0:
                    ret += "|  "*level+"|"+self.children[i].pretty_str(level+1)
                elif num_children==1 and self.children[0].val=="*":
                    return ret
                else:
                    ret += self.children[i].pretty_str(level+1) 
            elif level==0:
                if num_children>=2 and i!=0:
                    ret += "|\n"+"|"+self.children[i].pretty_str(level+1)
                else:
                    ret += self.children[i].pretty_str(level+1)
            
        return ret

    def get_max_level(self, level=0):
        if len(self.children) == 0:
            return level
        child_levels = [child.get_max_level(level=level + 1) for child in self.children]
        return max(child_levels)
class BuildLextree:
    def __init__(self, dictionary_file_path):
        self._convert_text_to_words(dictionary_file_path)
        self.root_node = LexNode('*')  
        self.root_node.property = 1
        total_words = len(self.words_list)
        word_lengths = [len(word) for word in self.words_list]
        self.longest_word_length = max(word_lengths)
        for index in range(total_words):
            self.words_list[index] = self.words_list[index].ljust(self.longest_word_length, '$')
        print(f"There are {len(self.words_list)} words in this dictionary.")
        
    def _convert_text_to_words(self, file_path):
        self.words_list = []
        with open(file_path, "r") as dictionary_file:
            lines = dictionary_file.readlines()
            for line in lines:
                self.words_list.append(line.strip('\n'))
        
    def add_lex_node(self, parent_node, child_node):
        
        assert isinstance(parent_node, LexNode) and isinstance(child_node, LexNode), "Both parent and child must be LexNode instances"
        parent_node.children.append(child_node)
    
    def construct_lexicon_tree(self):
        self._recursive_tree_build(self.root_node, self.words_list, 0)
        
    def _recursive_tree_build(self, current_node, remaining_words, current_index):
        if current_index == self.longest_word_length:
            print("Finished")
        words_ending_here = []
        for word in remaining_words:
            if (current_index + 1 < self.longest_word_length and word[current_index + 1] == '$') or current_index == self.longest_word_length - 1:
                end_node = LexNode(word[current_index])
                end_node.property = 2
                end_node.children.append(self.root_node)  # Add "*" at the end of the word
                self.add_lex_node(current_node, end_node)
                words_ending_here.append(word)
        for word in words_ending_here:
            remaining_words.remove(word)
        remaining_word_count = len(remaining_words)
        next_chars = [remaining_words[idx][current_index] for idx in range(remaining_word_count)]
        char_counts = Counter(next_chars)
        for char, count in char_counts.items():
            next_node = LexNode(char)
            self.add_lex_node(current_node, next_node)
            words_for_next_char = [word for word in remaining_words if word[current_index] == char]
            self._recursive_tree_build(next_node, words_for_next_char, current_index + 1)


In [ ]:
dic_file_path="dict_1.txt"

In [ ]:
buildlextree=BuildLextree(dic_file_path)
buildlextree.construct_lexicon_tree
lextree=buildlextree.root_node

There are 6250 words in this dictionary.


## Without beam

In [ ]:
import numpy as np
import copy
class SpellChecker:
    def __init__(self,beam,longest_word_lenth=45,tolerate_num_of_error=3):
        self.beam = beam
        self.lextree=None
        self.dist_fun=lambda *args: int(args[0] != args[1])
        self.longest_word_lenth=longest_word_lenth
        self.tolerate_num_of_error=tolerate_num_of_error

    def fit(self,lextree):
        self.lextree=lextree
        assert type(self.lextree) is LexNode
        self.nodes = []
        self.get_nodes(self.lextree)
        # get self.transitions
        self.transitions = {}
        self.get_chilren={}
        to_children={}
        n_nodes = len(self.nodes)
        self.word_ends = []
        # record the end idx of each words, therefore, at the end of the vertibe , we can get the costs of each word
        for i in range(n_nodes):
            n = self.nodes[i]
            #print(str(i)+"    "+n.val)
            if n.property == 2:
                self.word_ends.append(i)
            self.get_chilren[i]=[]
            # add transition if there is any. to get the parent node of current node
            if len(n.children) > 0:
                for child in n.children:
                    self.get_chilren[i].append(self.nodes.index(child))
                    self.transitions[self.nodes.index(child)] = i
        self.transitions[0]=0 # "start point repeating"
        
        
    def spell_check(self, text):
        # TODO: evaluate input
        pass
    def mute_nodes(self,r):
        self.muted_nodes[r]=1
        #mute it children
        to_mute_children=self.get_chilren[r]
        #print(to_mute_children)
        if len(to_mute_children)!=0:
            for child_idx in to_mute_children:
                if child_idx!=0:
                    self.mute_nodes(child_idx)
                
    def get_nodes(self, lexnode):
        lexnode.idx=len(self.nodes)
        self.nodes.append(lexnode)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            self.get_nodes(child)
    
    def text_vertibe(self,x,longest=True,get_string=True):
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        # initialize cost matrix
        n_cols = len(x)
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])

        # fill cost matrix
        for c in range(n_cols):
            for r in range(1,n_rows):
                if not self.muted_nodes[r]:
                    current_cost=self.dist_fun(self.nodes[r].val,x[c])
                    if c==0:#if it is the initialization.
                        if self.transitions[r]==0:
                            costs[r,c]=current_cost
                        else:
                            #if it is the middle nodes, then the cost is based on it previous node
                            costs[r,c]=costs[self.transitions[r],c]+current_cost
                    elif c==1:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        else:
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost
                    else:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        else:
                            #that means they are different
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c],#update from current state previous node
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost


                    if costs[r,c]>=self.tolerate_num_of_error+len(x):
                        self.mute_nodes(r)
            #print(costs[:25,c])
        
        
        #find the minimum cost at the end of each word
        costs_for_words=[]
        for end_idx in self.word_ends:
            costs_for_words.append(costs[end_idx,-1])
        #print(costs_for_words[0:40])
        
        min_cost=min(costs_for_words)
        #print(min_cost)
        #get the index of minimum cost
        min_idxs=np.where(costs_for_words==min_cost)[0]
        #print(min_idxs)
        if get_string:
            result=self.nodes_to_word(min_idxs,longest=longest)
            return result
        else:
            return min_idxs, min_cost
        
    #Methods used in Problem 2
    def nodes_to_word(self,nodes,longest=True):
        results=[]
        for idx in nodes:
            idx=self.word_ends[idx]
            #if the min cost is the real end node, then we can get a real word
            result=""
            while self.transitions[idx]!=0:
                result=self.nodes[idx].val+result
                #print(self.nodes[idx].val)
                idx=self.transitions[idx]
            result=self.nodes[idx].val+result
            results.append(result)
            #print(results)
        if longest:
            #return the longest word in the candidate
            res = max(results, key=len, default='')
            return res
        return results
    
    def nodes_to_sentence(self,nodes):
        sentence=""
        for idx in nodes:
            idx=self.word_ends[idx]
            #if the min cost is the real end node, then we can get a real word
            result=""
            while self.transitions[idx]!=0:
                result=self.nodes[idx].val+result
                #print(self.nodes[idx].val)
                idx=self.transitions[idx]
            result=self.nodes[idx].val+result
            sentence+=result+" "
        
        return  sentence
    
    def unsegment_text_vertibe(self,x,longest=True,n_best=3):
        self.final_nodes,self.final_costs=self._unsegment_text_vertibe(x)
        
        to_visual_nodes_links=[]
        
        small_number = heapq.nsmallest(n_best,self.final_costs) 
        small_index = []
        for t in small_number:
            index = self.final_costs.index(t)
            small_index.append(index)
            self.final_costs[index] = 0
            to_visual_nodes_links.append( self.final_nodes[index])
            
        visualized_sentence=[]
        for node_links in to_visual_nodes_links:
            visualized_result=self.nodes_to_sentence(node_links)
            print(visualized_result)
            visualized_sentence.append(visualized_result)
        return visualized_sentence
        
    def _unsegment_text_vertibe(self,x,threshold=10):
        returned_words=[]
        returned_costs=[]
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        
        # initialize cost matrix
        #n_cols is just the to check length
        if len(x)>self.tolerate_num_of_error+self.longest_word_lenth:
            n_cols=self.tolerate_num_of_error+self.longest_word_lenth
        else:
            n_cols=len(x)
    
        
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])
        
        # fill cost matrix
        for c in range(n_cols):
            print("current  c is "+str(c))
            for r in range(1,n_rows):
                if not self.muted_nodes[r]:
                    current_cost=self.dist_fun(self.nodes[r].val,x[c])
                    if c==0:#if it is the initialization.
                        if self.transitions[r]==0:
                            costs[r,c]=current_cost
                        else:
                            #if it is the middle nodes, then the cost is based on it previous node
                            costs[r,c]=costs[self.transitions[r],c]+current_cost
                    elif c==1:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        else:
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                     ] 
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost
                    else:
                        if current_cost==0:
                            to_check=[costs[r,c-1]+1, # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        else:
                            #that means they are different
                            to_check=[costs[r,c-1], # repeat the node from previous state
                                      costs[self.transitions[r],c-1],#get the node from previous state
                                      costs[self.transitions[r],c],#update from current state previous node
                                      costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                                     ]
                        nnn=np.argmin(to_check)
                        costs[r,c]=to_check[nnn]+current_cost


                    if costs[r,c]>=self.tolerate_num_of_error+len(x):
                        self.mute_nodes(r)
            #check whether there is already a new word complete
            #find the minimum cost at the end of each word
            if c!=0:
#                 print("current cost for all nodes is")
#                 print(costs[:,c])
#                 print(self.word_ends)
                costs_for_words=[]
                for end_idx in self.word_ends:
                    costs_for_words.append(costs[end_idx,c])
#                 
                min_cost=min(costs_for_words)
#                 
                min_idxs=np.where(costs_for_words==min_cost)[0]
                
                if len(x)<=n_cols:
                    print("there are less than {} charactor left!".format(n_cols))
                    print("minimum cost at this point is {}".format(min_cost))
                    rest_words=[]
                    if c<n_cols-1:
                        #check if we could push the rest letter to be new word
                        rest_string=x[c+1:]
                        rest_words,rest_min_cost=self.text_vertibe(rest_string,get_string=False)      
                          
                    for i in range(len(min_idxs)):
                        word_end_idx=min_idxs[i]
                        if len(rest_words)!=0:# if the rest characters could be a word, record the word and its cost
                            min_cost=min_cost+rest_min_cost
                            for rest_words_node in rest_words:
                                returned_words.append([word_end_idx,rest_words_node])
                                returned_costs.append(min_cost)
                        else: #else just add what I have now
                            returned_words.append([word_end_idx])
                            returned_costs.append(min_cost)
            
                          
                # elif, it is still in the window part of the input string
                else:
                    next_words,next_costs=self._unsegment_text_vertibe(x[c+1:])
                    for i in range(len(min_idxs)):
                        word_end_idx=min_idxs[i]
                        new_nodes=[]
                        new_costs=[]
                        for next_node_id in range(len(next_costs)):
                            next_node=next_words[next_node_id]
                            next_cost=next_costs[next_node_id]
#                             print("next cost is {}".format(next_cost)  )
#                             print("next words is {}".format(next_node)  )
                            current_cost=min_cost+next_cost
#                             print("current_cost is {}".format(current_cost))
                            if current_cost<threshold:
                                new_costs.append(current_cost)
                                new_nodes.append([word_end_idx]+next_node)
#                             print("newnodes!")
#                             print(new_nodes)
                        returned_words+=new_nodes
                        returned_costs+=new_costs
            print("after get the new nodes, my returned words would be")
            print(returned_words)
            print(returned_costs)
        
        return   returned_words,returned_costs               

## With beam

In [ ]:
import numpy as np
import copy
class SpellChecker:
    def __init__(self,beam,longest_word_lenth=45,tolerate_num_of_error=3):
        self.beam = beam
        self.lextree=None
        self.dist_fun=lambda *args: int(args[0] != args[1])
        self.longest_word_lenth=longest_word_lenth
        self.tolerate_num_of_error=tolerate_num_of_error

    def fit(self,lextree):
        self.lextree=lextree
        assert type(self.lextree) is LexNode
        self.nodes = []
        self.get_nodes(self.lextree)
        # get self.transitions
        self.transitions = {}
        self.get_children={}
        to_children={}
        n_nodes = len(self.nodes)
        self.word_ends = []
        # record the end idx of each words, therefore, at the end of the vertibe , we can get the costs of each word
        for i in range(n_nodes):
            n = self.nodes[i]
            #print(str(i)+"    "+n.val)
            if n.property == 2:
                self.word_ends.append(i)
            self.get_children[i]=[]
            # add transition if there is any. to get the parent node of current node
            if len(n.children) > 0:
                for child in n.children:
                    self.get_children[i].append(self.nodes.index(child))
                    self.transitions[self.nodes.index(child)] = i
        self.transitions[0]=0 # "start point repeating"
        
        
    def spell_check(self, text):
        # TODO: evaluate input
        pass
    def mute_nodes(self,r):
        self.muted_nodes[r]=1
        #mute it children
        to_mute_children=self.get_children[r]
        #print(to_mute_children)
        if len(to_mute_children)!=0:
            for child_idx in to_mute_children:
                if child_idx!=0:
                    self.mute_nodes(child_idx)
                
    def get_nodes(self, lexnode):
        lexnode.idx=len(self.nodes)
        self.nodes.append(lexnode)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            self.get_nodes(child)
    
    def text_vertibe(self,x,longest=True,get_string=True):
        #initialize the input x by adding "*" at the beginning of x. 
        x = '*' + x
        # initialize cost matrix
        n_cols = len(x)
        n_rows = len(self.nodes)
        costs = np.full((n_rows, n_cols), np.inf)
        costs[0,:]=range(n_cols)
        
        n_nodes = len(self.nodes)
        self.muted_nodes=np.zeros([n_nodes])

        # fill cost matrix
        for c in range(n_cols):
            self.recursion(self.lextree, c, x, costs)
            #print(costs[:25,c])
        
        
        #find the minimum cost at the end of each word
        costs_for_words=[]
        for end_idx in self.word_ends:
            costs_for_words.append(costs[end_idx,-1])
        #print(costs_for_words[0:40])
        
        min_cost=min(costs_for_words)
        #print(min_cost)
        #get the index of minimum cost
        min_idxs=np.where(costs_for_words==min_cost)[0]
        #print(min_idxs)
        if get_string:
            result=self.nodes_to_word(min_idxs,longest=longest)
            return result
        else:
            return min_idxs, min_cost
        
    #Beam
    def recursion(self, lexnode,c,x,costs):
        r=lexnode.idx
        #r=self.nodes.index(lexnode)
        if not self.muted_nodes[r]:
            current_cost=self.dist_fun(lexnode.val,x[c])
            if c==0:#if it is the initialization.
                if self.transitions[r]==0:
                    costs[r,c]=current_cost
                else:
                    #if it is the middle nodes, then the cost is based on it previous node
                    costs[r,c]=costs[self.transitions[r],c]+current_cost
            elif c==1:
                if current_cost==0:
                    to_check=[costs[r,c-1]+1, # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                             ] 
                else:
                    to_check=[costs[r,c-1], # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                             ] 
                nnn=np.argmin(to_check)
                costs[r,c]=to_check[nnn]+current_cost
            else:
                if current_cost==0:
                    to_check=[costs[r,c-1]+1, # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                              costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                             ]
                else:
                    #that means they are different
                    to_check=[costs[r,c-1], # repeat the node from previous state
                              costs[self.transitions[r],c-1],#get the node from previous state
                              costs[self.transitions[r],c],#update from current state previous node
                              costs[self.transitions[r],c-2]+1 # the c-1 is assumed as the insertion
                             ]
                nnn=np.argmin(to_check)
                costs[r,c]=to_check[nnn]+current_cost
            
            if costs[r,c]>=self.tolerate_num_of_error+len(x):
                    self.mute_nodes(r)

        #print(lexnode.val)
        if lexnode.property==2:
            return
        for child in lexnode.children:
            if costs[r,c]<=self.beam+1:
                self.recursion(child,c,x,costs)

# Initialize SpellChecker

In [ ]:
spellchecker=SpellChecker(3,longest_word_lenth=9)
#Fit lextree into spellchecker
spellchecker.fit(lextree)

In [ ]:
#Corrected typo with beam
typos_with_beam=['once upon a time wigle uncharacteristically as ing of uncharacteristically the uncharacteristically kamen to life at the foot of he uncharacteristically as a uncharacteristically he grew uncharacteristically and uncharacteristically big of uncharacteristically well to do and lived by a uncharacteristically of the uncharacteristically uncharacteristically in a uncharacteristically uncharacteristically now at that them uncharacteristically was a uncharacteristically uncharacteristically in the uncharacteristically the uncharacteristically uncharacteristically saw the uncharacteristically uncharacteristically of the uncharacteristically and she uncharacteristically a uncharacteristically to eke as uncharacteristically so she seda to her lord serb i uncharacteristically to feet the uncharacteristically of that uncharacteristically king of the uncharacteristically ', 'dodd life sadye the uncharacteristically i leed in the uncharacteristically and heed uncharacteristically on dry land how can we mach him ', 'dyk hua or by crock see uncharacteristically he met be kort if i uncharacteristically get heed i uncharacteristically die ', 'all rate uncharacteristically the uncharacteristically uncharacteristically ohr uncharacteristically uncharacteristically uncharacteristically i have a plan i wild give coo his cart to feet ', 'so when the uncharacteristically was uncharacteristically on the bank of the uncharacteristically after uncharacteristically a drink of water the uncharacteristically dromon near and seide sir uncharacteristically what do your live on bad uncharacteristically in this nolde uncharacteristically uncharacteristically on the uncharacteristically side of the uncharacteristically uncharacteristically is no end to the uncharacteristically uncharacteristically and uncharacteristically uncharacteristically wit uncharacteristically uncharacteristically as coney is it not bear to kroc uncharacteristically lande have allen uncharacteristically of uncharacteristically uncharacteristically to mate ', 'lore uncharacteristically the uncharacteristically uncharacteristically the uncharacteristically is uncharacteristically and uncharacteristically how shall i at uncharacteristically ', 'life coo want to gob i fill let ju sit upon my back and kady you over ', 'the uncharacteristically uncharacteristically am and uncharacteristically come here then seide the uncharacteristically up on myer back with coo and up the uncharacteristically uncharacteristically but when the uncharacteristically had swum a uncharacteristically wave he uncharacteristically the uncharacteristically under the uncharacteristically ', 'good uncharacteristically you are uncharacteristically me uncharacteristically uncharacteristically the uncharacteristically what is that or ', 'the uncharacteristically said you uncharacteristically i am uncharacteristically your out of uncharacteristically uncharacteristically uncharacteristically not a bit of it my wife has a uncharacteristically for uncharacteristically uncharacteristically and i uncharacteristically to eve it to ohr to mate ', 'uncharacteristically said the uncharacteristically it is nice of coo to tell me what if our cart uncharacteristically uncharacteristically us when we go uncharacteristically uncharacteristically the tried tops it wild be all uncharacteristically to uncharacteristically ', 'all where do you keep it asked the uncharacteristically ', 'the uncharacteristically uncharacteristically out a big tried with uncharacteristically of crye uncharacteristically uncharacteristically not far off side uncharacteristically he uncharacteristically are our uncharacteristically uncharacteristically on uncharacteristically fine tried ', 'if you uncharacteristically uncharacteristically me your uncharacteristically said the uncharacteristically then i uncharacteristically kill gob ', 'uncharacteristically meet to the uncharacteristically then and i all uncharacteristically it out to your ', 'the uncharacteristically uncharacteristically him to the uncharacteristically the uncharacteristically uncharacteristically off his back and uncharacteristically hupe the nigg tree sat uncharacteristically it oh uncharacteristically uncharacteristically uncharacteristically he you uncharacteristically that uncharacteristically were uncharacteristically that kept uncharacteristically uncharacteristically in a uncharacteristically you are a uncharacteristically and i have uncharacteristically you you may kept your uncharacteristically to uncharacteristically lore body is uncharacteristically but you have no uncharacteristically ', 'and uncharacteristically to uncharacteristically this uncharacteristically he uncharacteristically the uncharacteristically uncharacteristically ', 'uncharacteristically uncharacteristically uncharacteristically tops uncharacteristically the water uncharacteristically i see ', 'uncharacteristically of them i ant them not my nigg is uncharacteristically uncharacteristically for me ', 'uncharacteristically is yukon uncharacteristically uncharacteristically buts how uncharacteristically uncharacteristically is you with ', 'now go uncharacteristically ways serb uncharacteristically for i eve had the uncharacteristically how with ', 'the uncharacteristically uncharacteristically as said and uncharacteristically as if he had lost a uncharacteristically uncharacteristically of uncharacteristically ant uncharacteristically uncharacteristically to the uncharacteristically where he lived ']
corrected_withbeam=''
for x in typos_with_beam:
    corrected_withbeam+=x
    
corrected_withbeam

'once upon a time wigle uncharacteristically as ing of uncharacteristically the uncharacteristically kamen to life at the foot of he uncharacteristically as a uncharacteristically he grew uncharacteristically and uncharacteristically big of uncharacteristically well to do and lived by a uncharacteristically of the uncharacteristically uncharacteristically in a uncharacteristically uncharacteristically now at that them uncharacteristically was a uncharacteristically uncharacteristically in the uncharacteristically the uncharacteristically uncharacteristically saw the uncharacteristically uncharacteristically of the uncharacteristically and she uncharacteristically a uncharacteristically to eke as uncharacteristically so she seda to her lord serb i uncharacteristically to feet the uncharacteristically of that uncharacteristically king of the uncharacteristically dodd life sadye the uncharacteristically i leed in the uncharacteristically and heed uncharacteristically on dry land how can w

In [ ]:
#Correct stoty
f=open("storycorrect.txt")
lines = f.readlines()
temp=""
for line in lines:
    temp+=line
f.close()

temp

'Once upon a time, while Brahmadatta was king of Benares, the Bodhisatta came to life at the foot of the Himalayas as a monkey. He grew strong and sturdy, big of frame, well to do, and lived by a curve of the river Ganges in a forest haunt. Now at that time there was a crocodile dwelling in the Ganges. The crocodile\'s mate saw the great frame of the monkey, and she conceived a longing to eat his heart. So she said to her lord, "Sir, I desire to eat the heart of that great king of the monkeys!"\n\n"Good wife," said the crocodile, "I live in the water and he lives on dry land. How can we catch him?"\n\n"By hook or by crook," she replied, "he must be caught. If I don\'t get him, I shall die."\n\n"All right," answered the crocodile, consoling her, "don\'t trouble yourself. I have a plan. I will give you his heart to eat."\n\nSo when the Bodhisatta was sitting on the bank of the Ganges, after taking a drink of water, the crocodile drew near, and said, "Sir Monkey, why do you live on bad fr

# Accuracy

In [ ]:
def distance(x, y):
    return int(x!=y)

#Calculate the accuracy of the corrected result
def acc_checker1(inp, tem, distance):
    M, N=len(tem), len(inp)
    #print(M, N)
    cost=np.zeros((M, N))
    
    cost[0, 0] = distance(inp[0], tem[0])
    for i in range(1, M):
        cost[i, 0] = cost[i-1, 0] + distance(inp[0], tem[i])

    for j in range(1, N):
        cost[0, j] = cost[0, j-1] + distance(inp[j], tem[0])

    # Populate rest of cost matrix within window
    for i in range(1, N):
        for j in range(1, M):
        
            choices = cost[j, i-1], cost[j-1, i-1], cost[j-1, i]
            #print(i, j)
            cost[j, i] = min(choices) + distance(inp[i], tem[j])

    # Return DTW distance given window 
    return cost[-1, -1]

In [ ]:
#Cost between the corrected typos and the correct story
cost_nobeam=acc_checker1(corrected_withoutbeam, temp, distance)
cost_withbeam=acc_checker1(corrected_withbeam, temp, distance)
cost_nobeam, cost_withbeam

In [ ]:
#Accuracy of the corrected result
acc_nobeam=1-cost_nobeam/len(temp)
acc_withbeam=1-cost_withbeam/len(corrected_withbeam)
print('Accuracy without beam: ', acc_nobeam, 'Accuracy with beam 3: ', acc_withbeam)

Accuracy without beam:  0.8220258756705585 Accuracy with beam 3:  0.40364775239498896


#### The accuracy without beam width is about 82% while the accuracy with beam width 3 is only about 40%.